### ISSUE TITLE
Mighty Flip Champs DX, N+ (nplus), EDGE and possibly other minis - audio speed problem

### ISSUE THREAD

**Issue by benderscruffy** [hrydgard/ppsspp#9379]
music plays so low (you have to turn your speakers up to full ) and so slow
rename txt to zip
[ppsspplog.txt](https://github.com/hrydgard/ppsspp/files/818358/ppsspplog.txt)

**Comment by benderscruffy** [comment]
Ronnie O'Sullivans Snooker - NPEZ00411 
has the same problem on the title screen

**Comment by LunaMoo** [comment]
This is actually plagued by 2 issues:
1) yet another sceMp3 issue - it uses mp3 and ends up playing it at half speed / distorted, probably same issue as Quake II homebrew and pretty much all other games using Mp3 right now(well, all I tested at least:]),
2) the game sets the volume to 0xFE00 and we set that value same to two channels, this might need hardware testing, but I think there are only two possible solutions - volume should be either divided and set half for each channel or a safer option - clamped. Forcing max to 0x8000(which I took from other game) would be enough to fix this. Also worth noting value above 0x10000 completely breaks stuff, so volume one way or another probably should be clamped in case game sets wrong value.



 Unfortunately the easy to patch volume problem is useless to fix at this time, we have to figure out the sceMp3 issue mentioned above first since it's just nasty and provides worse experience than not having any music at all.:]

 I actually tried to look into that sceMp3 issue hence looking around for broken games, but I'm clueless and feeling like give up, I didn't even know a damn about mp3 format few hours ago, but now after reading quite a lot I'm conviced we parse mp3's fine, so the problem would be with decoder, the way we use it(FFMPEG I guess not the easiest thing to use;c) or possibly the output stream(pcm?) should be different for psp.:|

**Comment by unknownbrackets** [comment]
Hmm, we should add tests:
https://github.com/hrydgard/pspautotests/search?utf8=%E2%9C%93&q=sceVaudioOutputBlocking&type=

I think it's very likely the volume scale could be different - might be half as you say, for example.

If you're talking about the blocking in `__AudioEnqueue`, it's used by basically all audio in every game.  A lot of games block.  There ARE tests already for that code, it's a bit off but not majorly.

It's very possible that the data sent to `sceVaudioOutputBlocking` is in the wrong format (maybe even not interleaved, who knows?)  It could be playing wrong if there's double as much data or something.

-[Unknown]

**Comment by LunaMoo** [comment]
Progress! I think I figured out the issue with mp3:] soo far only tried it with this game and I hardcoded value just for a quick dirty check, but anyway I fixed the mp3 playback by setting:
```
chans[PSP_AUDIO_CHANNEL_OUTPUT2].sampleCount = 1152;
```
(inside sceVaudioOutputBlocking in addition to volume fix), for an mp3 which had a number of samples 1152:], we detect number of samples for mp3's so we're just not using it when we should?:] I'll try to get less hacky code and check more mp3's, but having a working mp3 will be nice:).

**Comment by unknownbrackets** [comment]
2304 is twice 1152.  This might be notable if it's an issue with stereo.  Maybe we just always divide by two in the vaudio func?

-[Unknown]

**Comment by LunaMoo** [comment]
Unfortunately as I listed above - mp3's with different format require different sampleCount and value set to it seems to always be 2304 and not just double ~ potentially set by completely different audio channel:|.

 I'm lost at where to get that sample count outside of decoder where it can be read from swrRet or frame_->nb_samples. Maybe on real hardware all audio decoding writes number of samples right before the stream or something ~ there is an empty place of 0xC bytes between some gradually rising counter and the pcm stream there which could potentially store some info about the stream, but that's just blind guessing.

 Also I noticed different apps/games with mp3 support can use different blocking syscalls, not just the vaudio one.

**Comment by unknownbrackets** [comment]
Hmm, okay, sorry, there's a lot of notes above.  Let me summarize a few things (feel free to correct any of the below):

 * `sceVaudioChReserve` is called to set the `sampleCount` value on the VAUDIO/OUTPUT2 channel.
 * VAUDIO shares the OUTPUT2 channel on a PSP - [as verified by tests](https://github.com/hrydgard/pspautotests/blob/master/tests/audio/sceaudio/reserve.expected#L165).
 * Games are statically calling `sceVaudioChReserve` with a sampleCount of 2304.  They are not getting 2304 from the result of an sceMp3 function call.
 * 2304 happens to be the number of 16-bit integers in a standard stereo mpeg layer 3 (mp3) frame (1152 left, 1152 right.)
 * Different mp3s have different timing issues depending on their frame size.
 * We don't know if sceMp3 is supposed to convert to a fixed number of samples (to normalize mp3 inputs for the game) when decoding.
 * There's some space that may have metadata from the mp3 written to it.

If the above is all true - then I think we'd need to write some tests to see how these functions actually behave on a real PSP.  Maybe even sceMp3 somehow directly changes the OUTPUT2 channel's sampleCount if it's already reserved?  Sounds super unlikely, but the easiest answer would be to decode an mp3 on a PSP and see what it spits out.

That said, the most probable thing - if games are hardcoding 2304 - is probably that sceMp3 is supposed to do some resampling on decode so that it always IS 2304.  But, just a guess really.

I know there are lots of "audio hack" issues with sceMp3 and I suspect there are bugs in PPSSPP's sceMp3.  sceVaudio is also suspect.  Note that the only mp3-related test is very shallow in what it tests:
https://github.com/hrydgard/pspautotests/blob/master/tests/audio/mp3/mp3test.expected

The output blocking code is at least better understood and I would think it's unlikely to be off by a factor of 2:
https://github.com/hrydgard/pspautotests/blob/master/tests/audio/sceaudio/output.expected

Additionally - a sample value of 0 is perfectly valid and samples should not be considered "null terminated".  Detecting zero as the end would definitely result in bugs.

-[Unknown]

**Comment by LunaMoo** [comment]
Yeah sorry for the amount of notes/edits, don't know much what I'm doing here and only looking into it since no one else was for a long while.:c

 About "hardcoded" sample count I also found other values used in other apps like 1024 in quake II(which seems to fail decoding mp3's completely even if they decode in other apps, but it seems a separate problem that will need more digging) and 4096 in "Light Mp3" which is a music player homebrew that has one really useful option to switch between using libmad and ME to play mp3's which makes some things clearer.
 When libmad is doing the decode it has same empty space before the stream as it would while using sceMp3, so I think that would point against my blind guess of any metadata stored there. But that's not all, libmad points to the pcm stream which has size of 4096 samples, however right after it is another stream of about 2570 samples and it uses sceAudioSRCOutputBlocking to point to both of them, one after another. When the same player with same mp3 uses sceMp3 it only get's one stream of 1152 samples, so quite a few times less:].

 That comparison would confirm your most probable guess that it has to be normalized - possibly can return more as with that libmad case, but should never return less as it always does now except maybe for cases where the game requests smaller number of samples than mp3 would give.:]

**Comment by LunaMoo** [comment]
Tried to look into fixing this and noticed, "Sound Speed hack" is pretty much doing it, but instead of doing it on decode as it should, it hacks around the problem when playing the pcm stream already, it also only deals with situation where sample count is exactly half of what it should be ~ which happens to also be the case of this game, but will not be the case for many others. I guess that Dead or Alive game it refers to is also affected by same issue overall, it was just popular enough to get a hack and the problem was never figured out:].

**Comment by hrydgard** [comment]
There is one game that I don't think uses sceMp3 but still has some kind of blocking sound timing issue, it's N+ (nplus) - so I wouldn't count out other kinds of issues entirely, though a bug in how we handle sceMp3 buffers seems kind of likely.

**Comment by LunaMoo** [comment]
I checked N+ and it seems to be broken only by sound volume which it sets to 0xF8000, dropping it down to 0x8000 as with this game here fixes it's sound effects and music completely.

**Comment by unknownbrackets** [comment]
`sceAudioOutputBlocking`/`sceAudioOutputPannedBlocking`/`sceAudioOutput` / etc. use 0xFFFF as the limit.

I don't remember the context now, but I *think* that some game was failing because it got an error code, triggering c38eece.  Tests do show that value is valid:

https://github.com/hrydgard/pspautotests/blob/7d9e54b31c1d5d4d72fdbebb1bc03ca1d8aaa05f/tests/audio/sceaudio/output.expected#L122

It's not clear what `0x10000` for Output2 actually does.  Tests aren't hooked into the actual speakers of the device, so can't tell us how the values operate.... it just tells us that the syscall accepts that volume value (while some other syscalls don't.)

Also note that this is treated as a signed value.  So 0xFFFF and 0x0001 are both "magnitude 1" in either direction, which is very quiet.  For the 8 regular channels, 0x7FFF and 0x8000 would be the largest magnitude available, loud.

Among the accepted Output2 range, presumably some of it is negative (maybe `0x80000` to `0xFFFFF`.)  Negative volume would mainly be used (I think) to cancel out sound in another channel.  I kinda doubt any PSP game does this when using the OUTPUT2/VAUDIO/SRC channel, though, but who knows.

See here for example:
http://manual.audacityteam.org/man/invert.html

My guess - from what I've seen - is that Output2/Vaudio/Src were added later, and represent the "newer" API.  Maybe they decided to provide finer control over the volume, so that you could do more than 100% volume (being `0x7FFF`), i.e. double it (being `0x10000` or so.)  In that case, we're probably not clamping properly after applying the left/right volume, or we're treating left/right as if they only have 16 bits when they have 20.

That theory would fit with the values 0xF8000 and 0xFE00 behaving as you described.  `0xF8000` would then be equivalent to -100% (so invert the wave, but otherwise 100% volume of the source samples), and `0x0FE00` would be equivalent to approximately +200% volume (well, 198.4%... basically double all source sample values.)

This sounds probably like a bug in the SSE volume handling optimizations.

-[Unknown]

**Comment by LunaMoo** [comment]
I think setting volume to unhearably low wouldn't have any sense from game design perspective, so if such values would be valid, maybe 0xFFFF shouldn't be an equivalent of 0x1 but of 0x7FFF.
 While testing other stuff I wasn't really overthinking it and just made myself a simple workaround for volume in __AudioEnqueue looking like this:
```
	while (chan.leftVolume > 0x8000) {
		chan.leftVolume >>= 1;
		chan.rightVolume >>= 1;
	}
```
So in both of those problematic games F8000 ends as 7C00, FE00 as 7F00, most games just sets volume to 8000 or lower so maybe not terribly bad.

 Also for the garbled/slowed down mp3's I completely removed "sound speed hack" and just left the check in AuCtx::AuDecode() as:
```
 while (sourcebuff.size() > 0 && outpcmbufsize < PCMBufSize) {
```
basically broken sound happens when buffer is underfilled so the whole hardcoded "repeat" there limiting it to 1(or 2 with "speed hack") might be the reason things break.

 Both of those solutions are just fitting the code to what games expect to work with not based on any tests, neither tested with huge amounts of games, but that's probably best I will be able to do here and leave it at that unless any of it is good enough for a pr.:|

 The only problem with mp3's I couldn't find any solution for yet is with Quake II which seems to end up with header-less mp3 streams that fail to decode, it seems to show some debug info on it's own, but it doesn't really say much other than it seeking to the start of the mp3 before playing it. then fails and closes.:]

**Comment by hrydgard** [comment]
I think you might be right about MP3, I'll look closer soon.

Though the notion of "headerless" MP3 confuses me, MP3 streams don't really have headers unless wrapped in a container like WAV, and can generally be decoded anyway since every frame has a full description of the encoding settings (bitrate etc) and it's quite easy to sync up to it, so header or not, it should work...

**Comment by LunaMoo** [comment]
Sorry I ment [frame header](https://www.mp3-tech.org/programmer/frame_header.html), which each mp3 frame should have.

With QII, here's a bit from log:
```
HLE\sceIo.cpp:1018 stdout: mp3_start_play(./MP3/08.mp3) @ 0
HLE\sceIo.cpp:1018 stdout: seek 0: 0/10813602
\sceMp3.cpp:182 sceMp3ReserveMp3Handle(0bfcee74)
HLE\sceIo.cpp:1018 stdout: Playing ./MP3/08.mp3
\sceMp3.cpp:314 sceMp3Init(0bfcee74)
\sceMp3.cpp:332 sceMp3Init(): channels=2, samplerate=44100Hz, bitrate=256kbps
\sceMp3.cpp:425 sceMp3GetSamplingRate(0BFCEE74)
\sceMp3.cpp:401 sceMp3GetMp3ChannelNum(0BFCEE74)
\sceMp3.cpp:389 sceMp3SetLoopNum(0BFCEE74, 0)
MediaEngine.cpp:86 FF: Header missing
SimpleAudioDec.cpp:198 Error decoding Audio frame (32768 bytes): -1094995529 (bebbb1b7)
HLE\sceIo.cpp:1018 stdout: mp3_end_play
HLE\sceIo.cpp:1018 stdout: Clear res
```
it'll repeat constantly since game tries without end, with some variations depending on mp3, but I tried the most standard common ones, even tried stripped of any ID3 tags as well as rarer formats, at most it might play a few frames.

Searching for those in-game log prints I found this open source DQuakePlus so maybe it's [Mp3](https://github.com/st1x51/DQuakePlus/blob/99ded70e28d75859a4e68868470e7d4726a35f53/psp/mp3.c) code can make it easier to figure out, but it could as well use different variant of this code as I saw some variations of it in the net, the code was potentially buggy and fixed from homebrew to homebrew.

 I think mp3's in this homebrew shouldn't at least have the buffer "underfill" problem since it reserves a rather small pcm buffer size of 1024 and most common mp3 types would have 1152.

**Comment by hrydgard** [comment]
There's something more complicated going on with the mp3 buffer size and "repeats" and stuff. Wipeout Pulse custom soundtrack support (place MP3s in memstick:MUSIC/WIPEOUT and they'll be selectable) plays tracks at double speed, skipping a lot, when the repeat check is eliminated. Not like it works better without though - with the old code they play too slowly, also with much skipping.

We're likely misinterpreting some buffer size or something.. the repeat stuff should probably go, though, indeed.

**Comment by LunaMoo** [comment]
Hmmm maybe there's some kind of upper limit, largest value reserved by game I saw - was set to 4096 samples. But the PCMBufSize itself can go to like 32768 if not higher. Hmmm maybe it's about https://github.com/jpcsp/jpcsp/blob/f6651937e64e563469de4f07e4f4a53ea8f21b24/src/jpcsp/HLE/modules/sceMp3.java#L303
where halfBufferSize = (bufferSize - reservedBufferSize) >> 1;

So in our case it might need to be an extra check like & outpcmbufsize < halfBuffer
where halfBuffer = (PCMBufSize - chans[PSP_AUDIO_CHANNEL_VAUDIO].sampleCount) / 2;

**Comment by LunaMoo** [comment]
As noted above removing the whole sound speed hack and leaving the check in in AuCtx::AuDecode() simpler as:
````
while (sourcebuff.size() > 0 && outpcmbufsize < PCMBufSize) {
````
solves the problem in all affected cases, using sound speed hack itself also helps, but only if the speed is exactly x2 times slower, so it's not really a good solution. Any against removal of this speedhack and solving the problem the way it was noted above? By now I tested it with many games since I use it all the time in my personal builds.

**Comment by hrydgard** [comment]
I haven't looked at it in detail, but it doesn't look wrong. Pull request it?

**Comment by LunaMoo** [comment]
Works fine for me, a bit annoying after a while, but works here. Althrough my copy is ID'ed as NPUZ00265 so maybe there's some difference(or simply corrupted file?).

**Comment by unknownbrackets** [comment]
NPUZ vs NPEZ means North America vs Europe.

-[Unknown]

**Comment by benderscruffy** [comment]
tested on my psp and it is quiet on the menu screen so it is the same as ppsspp
